runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -366,111 +489,137 @@ getClassSignature(J9VMThread *currentThread, J9Class * clazz)
-					sig[i] = '[';
+			if (sigLength <= stringBuffer->remaining()) {
+				for (U_32 i = 0; i < numDims; i++) {
+					stringBuffer->putCharAtCursor('[');
 				}
 
-				if (*name != '[') {
-					sig[i++] = 'L';
+				if ('[' != name[0]) {
+					stringBuffer->putCharAtCursor('L');
 				}
 
-				memcpy(sig+i, name, nameLength);
-				i += nameLength;
+				memcpy(stringBuffer->cursor, name, nameLength);
+				/* Adjust cursor to account for the memcpy. */
+				stringBuffer->advanceN(nameLength);
 
-				if (*name != '[') {
-					sig[i++] = ';';
+				if ('[' != name[0]) {
+					stringBuffer->putCharAtCursor(';');
 				}
-				sig[sigLength-1] = '\0';
+				result = true;
 			}
 		}
 	}
 
-	return sig;
+	return result;
 }
 
+/**
+ * Allocates a J9UTF8 signature buffer and places a method signature constructed from a MethodType into it.
+ * @param[in] currentThread The J9VMThread of the current thread
+ * @param[in] typeObject A MethodType object that contains parameter and return type information
+ * @return pointer to newly allocated and filled-in JUTF8 signature buffer
+ */
 static J9UTF8 *
-getJ9UTF8SignatureFromMethodType(J9VMThread *currentThread, j9object_t typeObject)
+getJ9UTF8SignatureFromMethodTypeWithMemAlloc(J9VMThread *currentThread, j9object_t typeObject)
 {
 	J9JavaVM *vm = currentThread->javaVM;
 	j9object_t ptypes = J9VMJAVALANGINVOKEMETHODTYPE_PTYPES(currentThread, typeObject);
 	U_32 numArgs = J9INDEXABLEOBJECT_SIZE(currentThread, ptypes);
-
-	J9UTF8* methodDescriptor = NULL;
-	char* cursor = NULL;
-	char* rSignature = NULL;
+	J9UTF8 *result = NULL;
 	UDATA signatureLength = 2; /* space for '(', ')' */
-
 	PORT_ACCESS_FROM_JAVAVM(vm);
 
-	char** signatures = (char**)j9mem_allocate_memory(numArgs * sizeof(char*), OMRMEM_CATEGORY_VM);
-	if (NULL == signatures) {
-		goto done;
-	}
-
-	memset(signatures, 0, numArgs * sizeof(char*));
-
+	/* Calculate total signature length, including all ptypes and rtype. */
 	for (U_32 i = 0; i < numArgs; i++) {
 		j9object_t pObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, ptypes, i);
 		J9Class *pclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, pObject);
-		signatures[i] = getClassSignature(currentThread, pclass);
-		if (NULL == signatures[i]) {
-			goto done;
-		}
-
-		signatureLength += strlen(signatures[i]);
+		signatureLength += getClassSignatureLength(currentThread, pclass);
 	}
+	j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
+	J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
+	signatureLength += getClassSignatureLength(currentThread, rclass);
 
-	{
-		/* Return type */
-		j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
-		J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
-		rSignature = getClassSignature(currentThread, rclass);
-		if (NULL == rSignature) {
-			goto done;
+	UDATA signatureUtf8Size = signatureLength + sizeof(J9UTF8) + 1 /* null-terminator */;
```

Comment:
embedding the comment before `;` makes the code flow discontiguous
```suggestion
	UDATA signatureUtf8Size = signatureLength + sizeof(J9UTF8) + 1; /* +1 for the null-terminator */
```